### PR TITLE
Test for the sphere heat kernel

### PR DESCRIPTION
**Pull request by NoemieJaquier**
This implements a test which verifies that the sphere heat kernel solves the heat equation. Additionally, it solves some casting issues in the hypersphere space which are relevant when deriving through the heat kernel parameters.

### CODE REVIEW

File: geometric_kernels/eigenfunctions.py [hunk 1]
```diff
@@ -171,7 +169,7 @@ def _filter_weights(self, weights: B.Numeric) -> B.Numeric:
         # TODO(VD) write check for this.
         i = 0
         for num in self.num_eigenfunctions_per_level:
-            weights_per_level.append(weights[i] * B.ones(1))
+            weights_per_level.append(weights[i] * B.cast(B.dtype(weights), B.ones(1)))
```
Comment (stoprightthere) [inline geometric_kernels/eigenfunctions.py:172]: You can just use `B.ones(B.dtype(weights), 1)`

File: tests/kernels/test_sphere_heat_kernel.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    """
+    ambient_dim = manifold.dim + 1
+    manifold_dim = manifold.dim
+    ambient_onb = np.eye(ambient_dim)
+
+    projected_onb = np.ones_like(ambient_onb)
+
+    for j in range(ambient_dim):
+        cur_vec = ambient_onb[j, :]
+        projected_onb[j, :] = manifold.to_tangent(cur_vec, base_point=x)
```
Comment (stoprightthere) [inline tests/kernels/test_sphere_heat_kernel.py:55]: Can it be vectorized? Although it's a minor point

File: geometric_kernels/spaces/hypersphere.py [hunk 3]
```diff
@@ -170,3 +170,12 @@ def get_eigenvalues(self, num: int) -> B.Numeric:
             eigenfunctions.num_eigenfunctions_per_level,
         )  # [num,]
         return B.reshape(eigenvalues, -1, 1)  # [num, 1]
+
+    def ehess2rhess(self, x, egrad, ehess, direction):
+        """
```
Comment (stoprightthere) [inline geometric_kernels/spaces/hypersphere.py:175]: Maybe we should cite something? E.g. https://sites.uclouvain.be/absil/2013-01/Weingarten_07PA_techrep.pdf or some other source.

File: tests/kernels/test_sphere_heat_kernel.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+from geometric_kernels.kernels.geometric_kernels import MaternKarhunenLoeveKernel
+
+
+_TRUNCATION_LEVEL = 10
+_NU = 2.5
+
+
+def manifold_laplacian(manifold, x, egrad, ehess):
+    r"""
+    Computes the manifold Laplacian of a given function at a given point x.
```
Comment (stoprightthere) [inline tests/kernels/test_sphere_heat_kernel.py:16]: This looks like a useful utility, maybe promote it to the core codebase as opposed to the test?

File: tests/kernels/test_sphere_heat_kernel.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        params["lengthscale"] = B.sqrt(2*t)
+        return kernel.K(params, state, x, y)
+
+    for t in ts:
+        for x in xs:
+            for y in ys:
+                # Compute the derivative of the kernel function wrt t
+                dfdt, _, _ = torch.autograd.grad(heat_kernel(t, x[None], y[None]), (t, x, y))
+                # Compute the Laplacian of the kernel on the manifold
+                egrad = lambda u: torch.autograd.grad(heat_kernel(t, u[None], y[None]), (t, u, y))[1]
```
Comment (stoprightthere) [inline tests/kernels/test_sphere_heat_kernel.py:98]: I see no point in writing a `def` here as the linter suggests. I think turning off the linter for these particular lines is fine.

**Comment by stoprightthere** [comment]
All in all, looks good to me! I suggest fixing linting errors (also see the comments).

File: tests/kernels/test_sphere_heat_kernel.py (hunk 4 above)
Comment (NoemieJaquier) [reply tests/kernels/test_sphere_heat_kernel.py:16]: Sure, that may be better! Where would you see it in the codebase?

File: geometric_kernels/eigenfunctions.py (hunk 1 above)
Comment (NoemieJaquier) [reply geometric_kernels/eigenfunctions.py]: Done, thanks :)

File: tests/kernels/test_sphere_heat_kernel.py (hunk 2 above)
Comment (NoemieJaquier) [reply tests/kernels/test_sphere_heat_kernel.py:55]: I am not sure to understand, do you mean done in a batch manner, or something like this?

File: tests/kernels/test_sphere_heat_kernel.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+import geometric_kernels.torch
+from geometric_kernels.spaces.hypersphere import Hypersphere
+from geometric_kernels.kernels.geometric_kernels import MaternKarhunenLoeveKernel
+
+
+_TRUNCATION_LEVEL = 10
+_NU = 2.5
+
+
+def manifold_laplacian(manifold, x, egrad, ehess):
```
Comment (lrozo) [inline tests/kernels/test_sphere_heat_kernel.py:14]: I suggest to add a reference to the equation of the manifold Laplacian.

File: tests/kernels/test_sphere_heat_kernel.py (hunk 2 above)
Comment (lrozo) [reply tests/kernels/test_sphere_heat_kernel.py:55]: What I Understood is that we may check if `manifold.to_tangent()` might handle an array of vectors, if so, then we may do this in a single shot.

**Review by lrozo — APPROVED** [review]
Nice PR, @NoemieJaquier ! Not really critical comments from my side.

File: tests/kernels/test_sphere_heat_kernel.py (hunk 2 above)
Comment (NoemieJaquier) [reply tests/kernels/test_sphere_heat_kernel.py]: We indeed can, I just updated this.

**Review by stoprightthere — APPROVED** [review]
Good PR, fine by me!

**Pull request merged**